index.html

```diff
@@ -1,390 +1,1451 @@
+          } finally {
+            reader.releaseLock();
+          }
+        }
+      </pre>
+
+      The inner loop will read chunks of data from the port until an error is
+      encountered, at which point the code in the "catch" block will be
+      executed. The outer loop handles recoverable errors such as parity check
+      failures by opening a new reader. Fatal errors will cause
+      {{SerialPort/readable}} to become `null` and the loop to end.
+
+      <p>
+      As long as the serial port is open it can continue to produce data and the
+      amount of data in each of the chunks returned by
+      {{ReadableStreamDefaultReader/read()}} will be essentially arbitrary based
+      on the timing of when it is called. It is up to the device and the code
+      communicating with it to decide what constitutes a complete message. For
+      example, a device might communicate with the host using ASCII-formatted
+      text where each message ends with a newline (or the sequence `"\r\n"`). A
+      pipeline of {{TransformStream}}s can be used to automatically convert the
+      {{Uint8Array}} chunks provided by {{SerialPort/readable}} into
+      {{DOMString}}s containing an entire line of text each.
+
+      <pre class="js">
+        class LineBreakTransformer {
+          constructor() {
+            this.container = '';
+          }
+
+          transform(chunk, controller) {
+            this.container += chunk;
+            const lines = this.container.split('\r\n');
+            this.container = lines.pop();
+            lines.forEach(line => controller.enqueue(line));
+          }
+
+          flush(controller) {
+            controller.enqueue(this.container);
+          }
+        }
+
+        const lineReader = port.readable
+            .pipeThrough(new TextDecoderStream())
+            .pipeThrough(new TransformStream(new LineBreakTransformer()))
+            .getReader();
+      </pre>
+
+      Some other ways of encoding message boundaries are to prefix each message
+      with its length or to wait a defined length of time before transmitting
+      the next message. Implementing a {{TransformStream}} for these types of
+      message boundaries is left as an exercise for the reader.
+    </aside>
+
+    The {{SerialPort/readable}} getter steps are:
+
+    <ol>
+      <li>
+        If [=this=].{{[[readable]]}} is not `null`, return
+        [=this=].{{[[readable]]}}.
+      <li>If [=this=].{{[[state]]}} is not `"opened"`, return `null`.
+      <li>If [=this=].{{[[readFatal]]}} is `true`, return `null`.
+      <li>Let |stream| be a [=new=] {{ReadableStream}}.
+      <li>
+        Let |pullAlgorithm| be the following steps, given
+        |controller:ReadableStreamDefaultController|:
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          <li>[=Resolve=] |promise| with `undefined`.
+          <li>
+            Return |promise| and run the following steps [=in parallel=]:
```

Comment:
I generally prefer to avoid doing things after returning. So I'd suggest this algorithm become:

1. Run the following steps in parallel:
   1. ...
2. Return a promise resolved with undefined.